### PR TITLE
Define webnn_wire.json and generate client code

### PR DESCRIPTION
**Pull request by fujunwei**

### CODE REVIEW

**Comment by fujunwei** [comment]
@huningxin MobileNet (maybe other examples) can work with wire layer now. PTAL, thanks

**Comment by fujunwei** [comment]
@huningxin PTAL, thanks.

File: webnn_wire.json [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      {"name": "object id", "type": "ObjectId" }
+    ],
+    "create graph builder": [
+      {"name": "context", "type": "ObjectId" },
+      {"name": "result", "type": "ObjectHandle", "handle_type": "graph builder" }
+    ]
+  },
+  "return commands": {
+    "context pop error scope callback": [
+      { "name": "context", "type": "ObjectHandle", "handle_type": "context" },
```
Comment (huningxin) [inline webnn_wire.json:76]: There are extra spaces after and before `{` and `}`.

File: webnn_wire.json [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      { "name": "request serial", "type": "uint64_t" },
+      { "name": "type", "type": "error type" },
+      { "name": "message", "type": "char", "annotation": "const*", "length": "strlen" }
+    ],
+    "graph compute result": [
+      { "name": "named outputs", "type": "ObjectHandle", "handle_type": "named outputs" },
```
Comment (huningxin) [inline webnn_wire.json:82]: ditto.

File: third_party/common/dawn/dawn_wire/server/ObjectStorage.h [hunk 3]
```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 The Dawn Authors
```
Comment (huningxin) [inline third_party/common/dawn/dawn_wire/server/ObjectStorage.h:1]: should we use this file in dawn?

File: src/webnn_wire/server/ServerContext.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+namespace webnn_wire { namespace server {
+
+    // void Server::OnUncapturedError(ObjectHandle context, MLErrorType type, const char* message) {
+    //     ReturnContextUncapturedErrorCallbackCmd cmd;
+    //     cmd.context = context;
+    //     cmd.type = type;
+    //     cmd.message = message;
+
+    //     SerializeCommand(cmd);
+    // }
```
Comment (huningxin) [inline src/webnn_wire/server/ServerContext.cpp:27]: If we don't need above commented lines, please remove them.

File: src/webnn_wire/client/ClientDoers.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    //         default:
+    //             return false;
+    //     }
+    //     if (device == nullptr) {
+    //         // The device might have been deleted or recreated so this isn't an error.
+    //         return true;
+    //     }
+    //     device->HandleError(errorType, message);
+    //     return true;
+    // }
```
Comment (huningxin) [inline src/webnn_wire/client/ClientDoers.cpp:39]: If we don't above commented lines, please remove them.

File: src/webnn_wire/WireServer.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    bool WireServer::InjectNamedOutputs(MLNamedOutputs namedOutputs,
+                                        uint32_t id,
+                                        uint32_t generation) {
+        return mImpl->InjectNamedOutputs(namedOutputs, id, generation);
+    }
+
+    // WGPUContext WireServer::GetContext(uint32_t id, uint32_t generation) {
+    //     return mImpl->GetContext(id, generation);
+    // }
```
Comment (huningxin) [inline src/webnn_wire/WireServer.cpp:58]: ditto.

File: src/webnn_native/ops/Constant.h [hunk 7]
```diff
@@ -33,7 +33,14 @@ namespace webnn_native { namespace op {
             mDescriptor.dimensions = mDimensions.data();
             mDescriptor.dimensionsCount = mDimensions.size();
             mDescriptor.type = desc->type;
+#if defined(WEBNN_ENABLE_WIRE)
+            // Prevent destroy from allocator memory after hanlding the command.
```
Comment (huningxin) [inline src/webnn_native/ops/Constant.h:37]: a typo: hanlding -> handling

File: src/webnn_native/ops/Constant.h [hunk 8]
```diff
@@ -33,7 +33,14 @@ namespace webnn_native { namespace op {
...
+            mBuffer = malloc(arrayBuffer->byteLength);
```
Comment (huningxin) [inline src/webnn_native/ops/Constant.h:38]: Where is the memory freed?

File: src/webnn_native/NamedInputs.h [hunk 9]
```diff
@@ -16,14 +16,59 @@
...
 #include <string>
 
-#include "webnn_native/NamedRecords.h"
+#include "common/Log.h"
 #include "webnn_native/webnn_platform.h"
 
 namespace webnn_native {
 
-    class NamedInputsBase : public NamedRecords<Input> {};
+    // TODO::Use WebGPU instead of ArrayBufferView
```
Comment (huningxin) [inline src/webnn_native/NamedInputs.h:27]: Use GPUBuffer?

File: src/utils/TerribleCommandBuffer.h [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+        size_t GetMaximumAllocationSize() const override;
+
+        void* GetCmdSpace(size_t size) override;
+        bool Flush() override;
+
+      private:
+        webnn_wire::CommandHandler* mHandler = nullptr;
+        size_t mOffset = 0;
+        char mBuffer[1000000];
```
Comment (huningxin) [inline src/utils/TerribleCommandBuffer.h:37]: Why is the buffer size (1000000) used?

File: build_overrides/webnn_features.gni [hunk 11]
```diff
@@ -31,4 +31,6 @@ declare_args() {
   # Enables the compilation of WebNN's Null backend
   # (required for unittests, obviously non-conformant)
   webnn_enable_null = true
+
+  webnn_enable_wire = false
```
Comment (huningxin) [inline build_overrides/webnn_features.gni:35]: should we turn on wire support by default?

File: examples/SampleUtils.cpp [hunk 12]
```diff
@@ -163,7 +265,11 @@ namespace utils {
     }
 
     ml::Graph Build(const ml::GraphBuilder& builder, const std::vector<NamedOperand>& outputs) {
+#if defined(WEBNN_ENABLE_WIRE)
```
Comment (huningxin) [inline examples/SampleUtils.cpp:268]: Should we move both paths (wire enabled or not) into `CreateCppNamedOperands`?

File: examples/SampleUtils.cpp [hunk 13]
```diff
@@ -163,7 +265,11 @@ namespace utils {
...
+        static ml::NamedOperands namedOperands = CreateCppNamedOperands();
```
Comment (huningxin) [inline examples/SampleUtils.cpp:269]: Why is the static variable used?

**Review by huningxin — COMMENTED** [review]
Great work! Thanks much @fujunwei .

File: third_party/common/dawn/dawn_wire/server/ObjectStorage.h (hunk 3 above)
Comment (fujunwei) [reply third_party/common/dawn/dawn_wire/server/ObjectStorage.h:1]: Yes, Most code are the same as dawn, but there are different namespace and `ContextInfo ` is used in webnn, but `DeviceInfo` for dawn in the file.

File: src/utils/TerribleCommandBuffer.h (hunk 10 above)
Comment (fujunwei) [reply src/utils/TerribleCommandBuffer.h:37]: The file is reused from dawn.

File: build_overrides/webnn_features.gni (hunk 11 above)
Comment (fujunwei) [reply build_overrides/webnn_features.gni:35]: Some tests are failed that need need more investigation but current status doesn't block chromium integration. I plan to go ahead to enable it in chromium.

File: examples/SampleUtils.cpp (hunk 12 above)
Comment (fujunwei) [reply examples/SampleUtils.cpp:268]: Yes, the flag should be passed with command line args in the future.

**Comment by fujunwei** [comment]
Done, thanks.

File: build_overrides/webnn_features.gni (hunk 11 above)
Comment (huningxin) [reply build_overrides/webnn_features.gni:35]: It makes sense. Before we merge this, could you please file an issue about the failures? So others may hlep fix them.

File: examples/SampleUtils.cpp (hunk 12 above)
Comment (huningxin) [reply examples/SampleUtils.cpp:292]: I mean move the conditional code into `CreateCppNamedOperands`, so sample could just call it without caring about wire or not.

File: examples/SampleUtils.cpp (hunk 13 above)
Comment (huningxin) [reply examples/SampleUtils.cpp]: What's the reason?

File: third_party/common/dawn/dawn_wire/server/ObjectStorage.h (hunk 3 above)
Comment (huningxin) [reply third_party/common/dawn/dawn_wire/server/ObjectStorage.h:1]: It makes sense to fork it for now. What's the path of the original dawn file? Could we track we will refactory it somewhere?

Comment (fujunwei) [reply third_party/common/dawn/dawn_wire/server/ObjectStorage.h:1]: Add annotation of refactoring in the file.

File: build_overrides/webnn_features.gni (hunk 11 above)
Comment (fujunwei) [reply build_overrides/webnn_features.gni:35]: File the issue https://github.com/otcshare/webnn-native/issues/927

**Comment by fujunwei** [comment]
Done, thanks.

File: examples/SampleUtils.cpp (hunk 13 above)
Comment (fujunwei) [reply examples/SampleUtils.cpp]: Removed the static type.

**Review by huningxin — APPROVED** [review]
LGTM, thanks @fujunwei , great work!

**Comment by fujunwei** [comment]
Thanks for review, merge it.

**Pull request merged**